gps_common/src/utm_odometry_node.cpp

```diff
@@ -3,245 +3,170 @@
+    odom_tf.child_frame_id = child_frame_id;
+    odom_tf.transform.translation.x = odom.pose.pose.position.x;
+    odom_tf.transform.translation.y = odom.pose.pose.position.y;
+    odom_tf.transform.translation.z = odom.pose.pose.position.z;
+
+    odom_tf.transform.rotation.x = odom.pose.pose.orientation.x;
+    odom_tf.transform.rotation.y = odom.pose.pose.orientation.y;
+    odom_tf.transform.rotation.z = odom.pose.pose.orientation.z;
+    odom_tf.transform.rotation.w = odom.pose.pose.orientation.w;
+
+    odom_tf_broadcaster.sendTransform(odom_tf);
   }
-}
 
+  nav_msgs::Odometry odom_world;
+  // keep twist and covariance matrix
+  odom_world = odom;
+  // replace pose
+  odom_world.pose.pose.position.x = position(0);
+  odom_world.pose.pose.position.y = position(1);
+  odom_world.pose.pose.position.z = position(2);
+  odom_world.pose.pose.orientation.w = orientation.w();
+  odom_world.pose.pose.orientation.x = orientation.x();
+  odom_world.pose.pose.orientation.y = orientation.y();
+  odom_world.pose.pose.orientation.z = orientation.z();
+  world_odom_pub.publish(odom_world);
+}
 
 int main (int argc, char **argv) {
   ros::init(argc, argv, "utm_odometry_node");
   ros::NodeHandle node;
   ros::NodeHandle priv_node("~");
-  bool use_ins = true;
-  odom_init = false; //initialization flag for odom.
-  publish_odom_tf = false;
-  broadcast_world_tf = false;
   priv_node.param<std::string>("frame_id", frame_id, "");
   priv_node.param<std::string>("child_frame_id", child_frame_id, "");
-  priv_node.param<double>("rot_covariance", rot_cov, 99999.0);
-
-  priv_node.param<std::string>("world_frame_id", world_frame_id, "");
-  priv_node.param<std::string>("world_child_frame_id", world_child_frame_id, "");
-
-
-  if(!priv_node.getParam("use_ins", use_ins))
-	  ROS_WARN_STREAM("No use_ins param provided, using default: "<<use_ins);
-
-  if(!priv_node.getParam("publish_odom_tf", publish_odom_tf))
-	  ROS_WARN_STREAM("No publish_odom_tf param provided, using default: "<<publish_odom_tf);
-
-  if(!priv_node.getParam("broadcast_world_tf",broadcast_world_tf))
-	  ROS_WARN_STREAM("No world_tf param provided, using default: "<<broadcast_world_tf);
-
-
-  if(use_ins)
-  {
-	  odom_pub = node.advertise<nav_msgs::Odometry>("/odom", 10);
-	  world_odom_pub = node.advertise<nav_msgs::Odometry>("/odom_world", 10);
-	  gps_source_sub = node.subscribe("/inspva", 10, insCallBack);
-  }
-  else
-  {
-	  odom_pub = node.advertise<nav_msgs::Odometry>("odom", 10);
-	  gps_source_sub = node.subscribe("fix", 10, callback);
+  priv_node.param<std::string>("world_frame_id", world_frame_id, "world");
+  priv_node.param<bool>("publish_odom_tf", publish_odom_tf, false);
+  priv_node.param<bool>("ignore_z", ignore_z, false);
+
+  // more than parameters, a service call would be better
+  // so that we could change the origin at runtime
+  geometry_msgs::PoseStamped origin;
+  priv_node.param<bool>("use_fixed_origin", use_fixed_origin, false);
```

Comment:
what about a function in chargte of setting up parameters, Moving to a service will be even easier after :)